yash-env/src/input/reporter.rs

```diff
@@ -0,0 +1,266 @@
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::job::fmt::Accumulator;
+use crate::option::{Interactive, Monitor, Off};
+use crate::Env;
+use async_trait::async_trait;
+use std::cell::RefCell;
+use yash_syntax::input::{Context, Input, Result};
+use yash_syntax::syntax::Fd;
+
+/// `Input` decorator that reports job status changes before reading a line
+///
+/// This decorator prints the status of jobs that have changed since the last
+/// report. The status is printed to the standard error before the input is read.
+/// This is done only if the [`Interactive`] and [`Monitor`] options are enabled.
+#[derive(Clone, Debug)]
+pub struct Reporter<'a, 'b, T: 'a> {
+    inner: T,
+    env: &'a RefCell<&'b mut Env>,
+}
+
+impl<'a, 'b, T> Reporter<'a, 'b, T> {
+    /// Creates a new `Reporter` decorator.
+    ///
+    /// The first argument is the inner `Input` that performs the actual input
+    /// operation. The second argument is the shell environment that contains
+    /// the shell option state and the system interface to print to the standard
+    /// error. It is wrapped in a `RefCell` so that it can be shared with other
+    /// decorators and the parser.
+    pub fn new(inner: T, env: &'a RefCell<&'b mut Env>) -> Self {
+        Self { inner, env }
+    }
+}
+
+#[async_trait(?Send)]
+impl<'a, 'b, T> Input for Reporter<'a, 'b, T>
+where
+    T: Input + 'a,
+{
+    #[allow(clippy::await_holding_refcell_ref)]
+    async fn next_line(&mut self, context: &Context) -> Result {
+        report(&mut self.env.borrow_mut()).await;
+        self.inner.next_line(context).await
+    }
+}
+
+async fn report(env: &mut Env) {
+    if env.options.get(Interactive) == Off || env.options.get(Monitor) == Off {
+        return;
+    }
+
+    let mut accumulator = Accumulator::new();
+    accumulator.current_job_index = env.jobs.current_job();
+    accumulator.previous_job_index = env.jobs.previous_job();
+    env.jobs
+        .iter()
+        .filter(|(_, job)| job.state_changed)
+        .for_each(|(index, job)| accumulator.add(index, job, &env.system));
+
+    if env
+        .system
+        .write_all(Fd::STDERR, accumulator.print.as_bytes())
+        .await
+        .is_ok()
+    {
+        for index in accumulator.indices_reported {
+            env.jobs.get_mut(index).unwrap().state_reported();
+        }
+    }
+}
```

Comment:
It is a good idea to report any errors that may occur. In this case, however, it is unlikely for the error reporting to succeed since we just failed to write to the standard error. I chose to silently discard the error.